freelist.go

```diff
@@ -13,22 +13,69 @@ type txPending struct {
 	alloctx          []txid // txids allocating the ids
 	lastReleaseBegin txid   // beginning txid of last matching releaseRange
 }
+type pidSet map[pgid]struct{}
 
 // freelist represents a list of all pages that are available for allocation.
 // It also tracks pages that have been freed but are still in use by open transactions.
 type freelist struct {
-	ids     []pgid              // all free and available free page ids.
-	allocs  map[pgid]txid       // mapping of txid that allocated a pgid.
-	pending map[txid]*txPending // mapping of soon-to-be free page ids by tx.
-	cache   map[pgid]bool       // fast lookup of all free and pending page ids.
+	ids         []pgid              // all free and available free page ids.
+	allocs      map[pgid]txid       // mapping of txid that allocated a pgid.
+	pending     map[txid]*txPending // mapping of soon-to-be free page ids by tx.
+	cache       map[pgid]bool       // fast lookup of all free and pending page ids.
+	freemaps    map[uint64]pidSet   // key is the size of continuous pages(span), value is a set which contains the starting pgids of same size
+	forwardMap  map[pgid]uint64     // key is start pgid, value is its span size
+	backwardMap map[pgid]uint64     // key is end pgid, value is its span size
 }
 
 // newFreelist returns an empty, initialized freelist.
 func newFreelist() *freelist {
 	return &freelist{
-		allocs:  make(map[pgid]txid),
-		pending: make(map[txid]*txPending),
-		cache:   make(map[pgid]bool),
+		allocs:      make(map[pgid]txid),
+		pending:     make(map[txid]*txPending),
+		cache:       make(map[pgid]bool),
+		freemaps:    make(map[uint64]pidSet),
+		forwardMap:  make(map[pgid]uint64),
+		backwardMap: make(map[pgid]uint64),
+	}
+}
+
+// initial from 0
+// pgids must be sorted
+func (f *freelist) init(pgids []pgid) {
+	f.freemaps = make(map[uint64]pidSet)
+	f.forwardMap = make(map[pgid]uint64)
+	f.backwardMap = make(map[pgid]uint64)
+	if len(pgids) == 0 {
+		return
+	}
+	if !sort.SliceIsSorted([]pgid(pgids), func(i, j int) bool { return pgids[i] < pgids[j] }) {
+		panic("pgids not sorted")
+	}
+
+	if len(pgids) == 1 {
```

Comment:
do we really need this special case handling?

Reply:
no need